### PR TITLE
[full-ci] [tests-only] Cache short-term artifacts to `cache` bucket

### PR DESCRIPTION
**Pull request by grgprarup**
## Description
In this PR, short-term artifacts are cached into the `cache` bucket and cached artifacts in the particular build are purged after the build is completed with status success or failure. 

The purge method cannot be called when the build is canceled so purging is done on the nightly build. 

Also, it clears empty directories which were created during the rebuilding cache and those not cleared by the build purge.

## Related Issue

- Fixes https://github.com/owncloud/QA/issues/757
- Fixes https://github.com/owncloud/QA/issues/758

## How Has This Been Tested?

- CI

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by ownclouders** [comment]
:boom: Acceptance test [<strong>cs3ApiTests-ocis</strong>](https://drone.owncloud.com/owncloud/ocis/14096/40/1) failed. Further test are cancelled...

**Comment by grgprarup** [comment]
Rebuild: cache uploaded to `cache` bucket 
`time="2022-08-10T14:50:18Z" level=info msg="file uploaded" bucket=cache`
https://drone.owncloud.com/owncloud/ocis/14343/3/5
https://drone.owncloud.com/owncloud/ocis/14343/4/3
https://drone.owncloud.com/owncloud/ocis/14343/4/4

Restore: cache restored using `cache` bucket
`time="2022-08-10T14:53:48Z" level=info msg="downloading file"`
https://drone.owncloud.com/owncloud/ocis/14343/44/2
https://drone.owncloud.com/owncloud/ocis/14343/44/5
https://drone.owncloud.com/owncloud/ocis/14343/44/6

Purge: cache purged
https://drone.owncloud.com/owncloud/ocis/14343/77/2
```
time="2022-08-10T15:21:44Z" level=info msg="found objects" bucket=cache count=3 key=owncloud/ocis/a436afb82d634dc721faaeced09e3a17a680192f-14343
time="2022-08-10T15:21:44Z" level=info msg="deleting object" bucket=cache key=owncloud/ocis/a436afb82d634dc721faaeced09e3a17a680192f-14343/ocis-binary-amd64_build_artifact_cache.tar
time="2022-08-10T15:21:44Z" level=info msg="deleting object" bucket=cache key=owncloud/ocis/a436afb82d634dc721faaeced09e3a17a680192f-14343/testing_app_build_artifact_cache.tar
time="2022-08-10T15:21:44Z" level=info msg="deleting object" bucket=cache key=owncloud/ocis/a436afb82d634dc721faaeced09e3a17a680192f-14343/testrunner_build_artifact_cache.tar
time="2022-08-10T15:21:44Z" level=info msg="Cache flushed"
```

**Comment by grgprarup** [comment]
After the merge of this PR:
TODO:
- [ ] old cache in bucket `owncloud` must be cleaned up

**Comment by SwikritiT** [comment]
> Rebuild: cache uploaded to `cache` bucket `time="2022-08-10T14:50:18Z" level=info msg="file uploaded" bucket=cache` https://drone.owncloud.com/owncloud/ocis/14343/3/5 https://drone.owncloud.com/owncloud/ocis/14343/4/3 https://drone.owncloud.com/owncloud/ocis/14343/4/4
> 
> Restore: cache restored using `cache` bucket `time="2022-08-10T14:53:48Z" level=info msg="downloading file"` https://drone.owncloud.com/owncloud/ocis/14343/44/2 https://drone.owncloud.com/owncloud/ocis/14343/44/5 https://drone.owncloud.com/owncloud/ocis/14343/44/6
> 
> Purge: cache purged https://drone.owncloud.com/owncloud/ocis/14343/77/2
> 
> ```
> time="2022-08-10T15:21:44Z" level=info msg="found objects" bucket=cache count=3 key=owncloud/ocis/a436afb82d634dc721faaeced09e3a17a680192f-14343
> time="2022-08-10T15:21:44Z" level=info msg="deleting object" bucket=cache key=owncloud/ocis/a436afb82d634dc721faaeced09e3a17a680192f-14343/ocis-binary-amd64_build_artifact_cache.tar
> time="2022-08-10T15:21:44Z" level=info msg="deleting object" bucket=cache key=owncloud/ocis/a436afb82d634dc721faaeced09e3a17a680192f-14343/testing_app_build_artifact_cache.tar
> time="2022-08-10T15:21:44Z" level=info msg="deleting object" bucket=cache key=owncloud/ocis/a436afb82d634dc721faaeced09e3a17a680192f-14343/testrunner_build_artifact_cache.tar
> time="2022-08-10T15:21:44Z" level=info msg="Cache flushed"
> ```

Yarn cache are also in `long-term` artifacts?

**Comment by grgprarup** [comment]
> > Rebuild: cache uploaded to `cache` bucket `time="2022-08-10T14:50:18Z" level=info msg="file uploaded" bucket=cache` https://drone.owncloud.com/owncloud/ocis/14343/3/5 https://drone.owncloud.com/owncloud/ocis/14343/4/3 https://drone.owncloud.com/owncloud/ocis/14343/4/4
> > Restore: cache restored using `cache` bucket `time="2022-08-10T14:53:48Z" level=info msg="downloading file"` https://drone.owncloud.com/owncloud/ocis/14343/44/2 https://drone.owncloud.com/owncloud/ocis/14343/44/5 https://drone.owncloud.com/owncloud/ocis/14343/44/6
> > Purge: cache purged https://drone.owncloud.com/owncloud/ocis/14343/77/2
> > ```
> > time="2022-08-10T15:21:44Z" level=info msg="found objects" bucket=cache count=3 key=owncloud/ocis/a436afb82d634dc721faaeced09e3a17a680192f-14343
> > time="2022-08-10T15:21:44Z" level=info msg="deleting object" bucket=cache key=owncloud/ocis/a436afb82d634dc721faaeced09e3a17a680192f-14343/ocis-binary-amd64_build_artifact_cache.tar
> > time="2022-08-10T15:21:44Z" level=info msg="deleting object" bucket=cache key=owncloud/ocis/a436afb82d634dc721faaeced09e3a17a680192f-14343/testing_app_build_artifact_cache.tar
> > time="2022-08-10T15:21:44Z" level=info msg="deleting object" bucket=cache key=owncloud/ocis/a436afb82d634dc721faaeced09e3a17a680192f-14343/testrunner_build_artifact_cache.tar
> > time="2022-08-10T15:21:44Z" level=info msg="Cache flushed"
> > ```
> 
> Yarn cache is also in `long-term` artifacts?

@SwikritiT Yeah. The cache of the web is used as a long-term cache, and the yarn cache for web e2e and acceptance goes under the web cache. Which are implemented by this pr https://github.com/owncloud/ocis/pull/4309

**Review by individual-it — CHANGES_REQUESTED** [review]

**Review by kiranparajuli589 — COMMENTED** [review]
other than the existing comments, lgtm :+1:

File: .drone.star [hunk 1]
```diff
@@ -2234,12 +2224,19 @@ def genericCachePurge(name, cache_key):
...
     if action == "rebuild" or action == "restore":
-        return genericCache(name, action, [path], cache_key)
+        name = "%s_build_artifact_cache" % (name)
+        return genericCache(name, action, [path], cache_path)
+
     if action == "purge":
-        return genericCachePurge(name, cache_key)
+        # cached items which are not purged due to build cancel are purged on nightly
+        if ctx.build.event == "cron":
+            flush_path = "%s/%s" % ("cache", ctx.repo.slug)
```
Comment (phil-davis) [inline .drone.star:2236]: I there is an ordinary build that happens to be running when the nightly cron build runs, then what is going to happen? The cache of that ordinary PR build might unexpectedly disappear.

We could set the `flush_age` (line 2209) to `1` day - "random rubbish" will sit for an extra day before being cleaned up, but we avoid accidentally breaking a running PR  build.

**Comment by phil-davis** [comment]
`This branch is [2 commits ahead](https://github.com/owncloud/ocis/compare/master...check-flush-cache), [31 commits behind](https://github.com/owncloud/ocis/compare/check-flush-cache...master) master.`

Please rebase when you address the comments, so that we get up-to-date CI.

File: .drone.star [hunk 2]
```diff
@@ -30,7 +30,7 @@ PLUGINS_GITHUB_RELEASE = "plugins/github-release:1"
 PLUGINS_GIT_ACTION = "plugins/git-action:1"
 PLUGINS_MANIFEST = "plugins/manifest:1"
 PLUGINS_S3 = "plugins/s3:latest"
-PLUGINS_S3_CACHE = "plugins/s3-cache:1"
+PLUGINS_S3_CACHE = "plugins/s3-cache:latest"
```
[the remark this answers is not in the document]
Comment (grgprarup) [reply .drone.star]: This plugin has images `plugins/s3-cache:1` and `plugins/s3-cache:1.4` which are 3 years old, there are no other latest specific versions except `plugins/s3-cache:latest` which is 17 month old image. 
plugins_image
Versions 1 and 1.4 are similar with no major changes. The latest tag version has a better output log than versions 1 and 1.4. 
Which version should we use 1 or the latest? @individual-it

**Review by phil-davis — APPROVED** [review]

**Pull request merged**